core/runtime/common/runtime_instances_pool.cpp

```diff
@@ -74,63 +74,60 @@ namespace kagome::runtime {
   outcome::result<std::shared_ptr<ModuleInstance>, CompilationError>
   RuntimeInstancesPool::instantiateFromCode(const CodeHash &code_hash,
                                             common::BufferView code_zstd) {
-    std::unique_lock lock{pools_mtx_};
-    auto pool_opt = pools_.get(code_hash);
-
-    if (!pool_opt) {
-      lock.unlock();
-      if (auto future = getFutureCompiledModule(code_hash)) {
-        lock.lock();
-        pool_opt = pools_.get(code_hash);
+    std::optional<std::reference_wrapper<InstancePool>> pool_opt{};
+    {
+      std::unique_lock lock{pools_mtx_};
+      pool_opt = pools_.get(code_hash);
+    }
+    while (!pool_opt) {
+      std::promise<CompilationResult> promise;
```

Comment:
[Don't create promise when not needed](https://github.com/qdrvm/kagome/compare/fix/instance-pool-2).